src/tracer.cc

```diff
@@ -374,25 +369,192 @@ TraceManager::Trace::InitTracer(
+  // to keep it alive for the duration of the request.
+  otel_context_ = opentelemetry::context::Context({kRootSpan, root_span});
 }
 
-void
-TraceManager::Trace::InitSpan(const otel_common::SystemTimestamp& timestamp_ns)
+opentelemetry::nostd::shared_ptr<otel_trace_api::Span>
+TraceManager::Trace::StartSpan(
+    std::string display_name, const uint64_t& raw_timestamp_ns,
+    bool is_root_span, std::string parent_span_key)
 {
   otel_trace_api::StartSpanOptions options;
-  options.kind = otel_trace_api::SpanKind::kServer;  // server
-  options.start_system_time = timestamp_ns;
-  // [FIXME] think about names
-  trace_span_ =
-      provider_->GetTracer("triton-server")->StartSpan("InferRequest", options);
+  options.kind = otel_trace_api::SpanKind::kServer;
+  options.start_system_time =
+      time_offset_ + std::chrono::nanoseconds{raw_timestamp_ns};
+  options.start_steady_time =
+      otel_common::SteadyTimestamp{std::chrono::nanoseconds{raw_timestamp_ns}};
+
+  // If the new span is a child span, we need to retrieve its parent from
+  // the context and provide it through StartSpanOptions to the child span
+  if (!is_root_span && otel_context_.HasKey(parent_span_key)) {
+    auto parent_span = opentelemetry::nostd::get<
+        opentelemetry::nostd::shared_ptr<otel_trace_api::Span>>(
+        otel_context_.GetValue(parent_span_key));
+    options.parent = parent_span->GetContext();
+  }
+  return provider_->GetTracer(kTritonTracer)->StartSpan(display_name, options);
+}
+
+void
+TraceManager::Trace::EndSpan(std::string span_key)
+{
+  auto timestamp_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(
+                          std::chrono::steady_clock::now().time_since_epoch())
+                          .count();
+  EndSpan(span_key, timestamp_ns);
 }
 
+
 void
-TraceManager::Trace::EndSpan()
+TraceManager::Trace::EndSpan(
+    std::string span_key, const uint64_t& raw_timestamp_ns)
 {
-  if (trace_span_ != nullptr) {
-    trace_span_->End();
+  if (otel_context_.HasKey(span_key)) {
+    auto span = opentelemetry::nostd::get<
+        opentelemetry::nostd::shared_ptr<otel_trace_api::Span>>(
+        otel_context_.GetValue(span_key));
+
+    if (span == nullptr) {
+      return;
+    }
+
+    otel_trace_api::EndSpanOptions end_options;
+    end_options.end_steady_time = otel_common::SteadyTimestamp{
+        std::chrono::nanoseconds{raw_timestamp_ns}};
+    span->End(end_options);
+  }
+}
+
+void
+TraceManager::Trace::ReportToOpenTelemetry(
+    TRITONSERVER_InferenceTrace* trace,
+    TRITONSERVER_InferenceTraceActivity activity, uint64_t timestamp_ns)
+{
+  uint64_t id;
+  LOG_TRITONSERVER_ERROR(
+      TRITONSERVER_InferenceTraceId(trace, &id), "getting trace id");
+
+  auto current_span_key = GetSpanKeyForActivity(activity, id);
+  if (current_span_key.empty()) {
+    return;
+  }
+
+  MaybeStartSpan(current_span_key, trace, activity, timestamp_ns, id);
```

Comment:
The "Maybe" logic is not completely captured in this function. 

I would propose creating a new wrapper function for calling an AddEvent including L450-L459. Based on the activity field, the function can create a span or not.